### ISSUE TITLE
Retention timer not work

### ISSUE THREAD

**Issue by SmartHome2021** [eduardolat/pgbackweb#50]
Hi, I'm having some problems with auto-deleting old archives. 
During the setup process, I experimented with the time of the auto-delete value, it seems that this does not affect anything.

At the moment, I have the following parameters set:
cloud backup - 30 days
local backup - 3 days

Despite this, local backups have been stored since 19.09, for 19 days now. 

I periodically delete them myself, but this is not the experience I would like to have

I will be glad of any advice. Thanks
![backup](https://github.com/user-attachments/assets/f9c4e6ae-43e6-4885-98d8-58d33327bab5)
![wault](https://github.com/user-attachments/assets/6a2746e8-f34e-4246-be44-ea4612672761)

**Comment by Stitch10925** [comment]
I am seeing the same problem. I have a fairly big database of around 3 GB that holds non-crucial data. As such I have set the retention to 2 days. However, the backup folder for the database has now grown to 21 GB. When looking into the folders I can indeed see that some old versions are not being deleted.

**Comment by SmartHome2021** [comment]
Is the cron of the system on which the image is deployed used to process the schedule, or is there another implementation? 
I'm asking because the system cron is empty.

**Comment by Stitch10925** [comment]
I also have a backup in the database of which the file does not exist anymore (because I changed the file structure). When the cron job runs it seems to fail on that one file and stop, so other backups from other databases are not cleaned up.

I also cannot manually delete the file, since I get the error that the file does not exist anymore. There seems to be no way to delete the database entry. 

Why not simply delete the database entry when a "delete" is tried but the file does not exist anymore?

**Comment by SmartHome2021** [comment]
> I also have a backup in the database of which the file does not exist anymore (because I changed the file structure). When the cron job runs it seems to fail on that one file and stop, so other backups from other databases are not cleaned up.
> 
> I also cannot manually delete the file, since I get the error that the file does not exist anymore. There seems to be no way to delete the database entry.
> 
> Why not simply delete the database entry when a "delete" is tried but the file does not exist anymore?

Try to create a text file with the extension using the path specified in the error backup_name.zip . In theory, this will be enough for the system to see the file and send it. 

Unfortunately, the situation is more complicated for me, I made a mistake in the installation and I have the first phantom backup stored on the path :/backups/backups/billing/<destination-directory>/<YYYY>/<MM>/<DD>/dump-<random-suffix>.zip/2024/09/17/dump-20240917-142336-01bd3366-3296-438d-958d-db0f36152afe.zip

The < symbol causes an error and the system cannot delete the file, even though I was late all the way. 

Such characters should be escaped with the \ character, then everything works correctly. 


I hope this error will be noticed by the developer and he will tell you what to do

**Comment by SmartHome2021** [comment]
> I also have a backup in the database of which the file does not exist anymore (because I changed the file structure). When the cron job runs it seems to fail on that one file and stop, so other backups from other databases are not cleaned up.
> 
> I also cannot manually delete the file, since I get the error that the file does not exist anymore. There seems to be no way to delete the database entry.
> 
> Why not simply delete the database entry when a "delete" is tried but the file does not exist anymore?

I have solved this problem, and I am giving you a solution.

1. Connect via ssh to the server on which docker is deployed
2. run sudo docker ps to display the list of containers
3. copy the container_id of the container that is responsible for the database (image postgres:*)
4. enter the sudo docker exec -it command (your copied id) bash
5. psql -U postgres
6. SELECT datname FROM pg_database;
7. \connect pgbackweb
8. \dt
9. SELECT * FROM executions;
10. we look for problematic entries and copy the path value for each
11. delete bad entries with the DELETE FROM executions WHERE path ='(copied path)' command

It looks like this for me : DELETE FROM executions WHERE path = 'backups/billing/<destination-directory>/<YYYY>/<MM>/<DD>/dump-<random-suffix>.zip/2024/09/17/dump-20240917-142336-01bd3366-3296-438d-958d-db0f36152afe.zip'

PROFIT! There is no need to reboot anything, we wait 10 minutes and open the catalog. The old versions have been deleted.

**Comment by TristisOris** [comment]
But retention still not working. I have 3 months old dumps, with 7days rule.

**Comment by eduardolat** [comment]
let me take a look

**Comment by eduardolat** [comment]
@Stitch10925 Can you please verify in your file system if the files in the logs exists?

**Comment by Stitch10925** [comment]
@eduardolat 
Yes, the files exist. When I manually delete them, the error goes away.

**Comment by Stitch10925** [comment]
Maybe to give a bit more context, at least in my case:

I use an NFS volume mount to my NAS to mount the directory where the backups are stored. Writing always seems to work, but removing seems to be an issue, even though manually deleting the files has never been a problem.

Hmm, I just realized, maybe something is changing the permissions of the files causing issues with the deletion, I will have to check that this weekend 🤔

**Comment by TristisOris** [comment]
Still not fixed.

```
2025-10-08T11:30:00.162905509+03:00 INFO msg=all databases tested

2025-10-08T11:40:00.001511961+03:00 INFO msg=old sessions deleted

2025-10-08T11:40:00.002390903+03:00 INFO msg=all destinations tested

2025-10-08T11:40:00.017484087+03:00 ERROR msg=error soft deleting expired executions error=failed to delete file /backups/backups/2024/12/28/dump-20241228-123943-aa3714e9-dfe8-4382-8d06-d2dc9d0b2fd3.zip: remove /backups/backups/2024/12/28/dump-20241228-123943-aa3714e9-dfe8-4382-8d06-d2dc9d0b2fd3.zip: no such file or directory id=17a1aa65-5342-4ed4-a669-b96b5288ed38

2025-10-08T11:40:00.124262482+03:00 INFO msg=all databases tested
```

this log repeated for 100 times. No attempts to remove any other date. I removed old backups manually, but db is not cleared.

**Comment by TristisOris** [comment]
<img width="229" height="311" alt="Image" src="https://github.com/user-attachments/assets/56641840-e7d7-4ca3-a330-d48bb49edc5f" />

empty folders not removed with files.

**Comment by BartschLabs** [comment]
I also have empty folders remaining after the files are deleted.

**Comment by zitom4** [comment]
Hello, I had the same problem. Apparently, the process that deletes historical files stops if it doesn’t find a file in the directory structure, just as (@TristisOris) mentioned in their message. In my case, I changed the directory structure where I stored the backups, and there are files that pgbackweb copied that no longer exist in the paths where they were stored.

The scheduled task that makes backups stores each backup in the "executions" table, along with its date, status, and path. The deletion script goes through this same table.

To temporarily fix this, we can connect to the Postgres container, access the pgbackweb database, and query the "executions" table. This table contains all the backup executions. To prevent the script from trying to delete a backup, it is enough to mark it as "deleted" and update the "delete_at" field.

In my case, I analyzed the backups by the date of the change and updated the ones I knew didn’t exist to status="deleted", updated_at="<date>", deleted_at="<date>".

I hope it helps.

**Comment by vutsalsinghal** [comment]
@eduardolat i'm facing issue with pgbackweb not adhering to retention settings as well. I have B2 as my S3 Destination and since I started using pgbackweb 3 months ago, none of the backups have been deleted even though retention is set to 10days.
